package NAlice;

option go_package = "a.yandex-team.ru/alice/megamind/protos/common";
option java_package = "ru.yandex.alice.megamind.protos.common";
option java_outer_classname = "Events";

import "alice/library/censor/protos/extension.proto";

import "mapreduce/yt/interface/protos/extension.proto";

import "google/protobuf/struct.proto";

enum EEventType {
    voice_input = 1;
    text_input = 2;
    suggested_input = 3;
    image_input = 4;
    music_input = 5;
    server_action = 6;
};

message TAsrResult {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    message TWord {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional string Value = 1 [json_name = "value", (NYT.column_name) = "value", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE];
        optional double Confidence = 2 [json_name = "confidence", (NYT.column_name) = "confidence"];
    };

    optional string Utterance = 1 [json_name = "utterance", (NYT.column_name) = "utterance", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE, deprecated = true];
    optional double Confidence = 2 [json_name = "confidence", (NYT.column_name) = "confidence"];
    repeated TWord Words = 3 [json_name = "words", (NYT.column_name) = "words"];
    optional string Normalized = 4 [json_name = "normalized", (NYT.column_name) = "normalized", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE];
};

message TImageInputData {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    optional string ImgUrl = 1 [json_name = "img_url", (NYT.column_name) = "img_url", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE];
};

message TMusicResult {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    message TData {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional google.protobuf.Struct Match = 1 [json_name = "match", (NYT.column_name) = "match", (NYT.flags) = SERIALIZATION_PROTOBUF]; // большой payload от музыки
        reserved 2;
        optional string RecognitionId = 3 [json_name = "recognition-id", (NYT.column_name) = "recognition_id"]; // according to https://a.yandex-team.ru/arc/trunk/arcadia/alice/vins/api/vins_api/speechkit/schemas.py?rev=5421968#L290
        optional string Engine = 4 [json_name = "engine", (NYT.column_name) = "engine"];
        optional string Url = 5 [json_name = "url", (NYT.column_name) = "url"];
    };

    optional TData Data = 1 [json_name = "data", (NYT.column_name) = "data"];
    optional string Result = 2 [json_name = "result", (NYT.column_name) = "result"];
    optional string ErrorText = 3 [json_name = "error_text", (NYT.column_name) = "error_text"];

};

message TBiometryScoring {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    message TScore {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional string UserId = 1 [json_name = "user_id", (NYT.column_name) = "user_id"];
        optional double Score = 2 [json_name = "score", (NYT.column_name) = "score"];
    }

    message TScoreWithMode {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional string Mode = 1 [json_name = "mode", (NYT.column_name) = "mode"];
        repeated TScore Scores = 2 [json_name = "scores", (NYT.column_name) = "scores"];
    }

    repeated TScore Scores = 1 [json_name = "scores", (NYT.column_name) = "scores"];
    repeated TScoreWithMode ScoresWithMode = 2 [json_name = "scores_with_mode", (NYT.column_name) = "scores_with_mode"]; // todo: wtf? for backward compatibility

    optional string Status = 3 [json_name = "status", (NYT.column_name) = "status"];
    optional string RequestId = 4 [json_name = "request_id", (NYT.column_name) = "request_id"];
    optional string GroupId = 5 [json_name = "group_id", (NYT.column_name) = "group_id"];
    optional int64 PartialNumber = 6 [json_name = "partial_number", (NYT.column_name) = "partial_number"];
}

message TBiometryClassification {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    message TClassificationResult {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional string ClassName = 1 [json_name = "classname", (NYT.column_name) = "classname"];
        optional double Confidence  = 2 [json_name = "confidence", (NYT.column_name) = "confidence"];
        optional string Tag  = 3 [json_name = "tag", (NYT.column_name) = "tag"];
    }

    message TClassificationSimple {
        option (NYT.default_field_flags) = SERIALIZATION_YT;

        optional string ClassName = 1 [json_name = "classname", (NYT.column_name) = "classname"];
        optional string Tag = 2 [json_name = "tag", (NYT.column_name) = "tag"];
    }

    optional string Status = 1 [json_name = "status", (NYT.column_name) = "status"];
    repeated TClassificationResult Scores = 2 [json_name = "scores", (NYT.column_name) = "scores"];
    repeated TClassificationSimple Simple = 3 [json_name = "simple", (NYT.column_name) = "simple"];
    optional int64 PartialNumber = 4 [json_name = "partial_number", (NYT.column_name) = "partial_number"];
}

message TEvent {
    option (NYT.default_field_flags) = SERIALIZATION_YT;

    optional EEventType Type = 1 [json_name = "type", (NYT.column_name) = "type"];

    // voice_input
    optional uint32 HypothesisNumber = 2 [json_name = "hypothesis_number", (NYT.column_name) = "hypothesis_number"];
    optional bool EndOfUtterance = 3 [json_name = "end_of_utterance", (NYT.column_name) = "end_of_utterance"];
    repeated TAsrResult AsrResult = 4 [json_name = "asr_result", (NYT.column_name) = "asr_result"];
    optional uint32 OriginalZeroAsrHypothesisIndex = 15 [json_name = "original_zero_asr_hypothesis_index", (NYT.column_name) = "original_zero_asr_hypothesis_index"];

    // text_input, suggest_input
    optional string Text = 5 [json_name = "text", (NYT.column_name) = "text", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE];

    // image_input
    // NOTE(g-kostin): In real speechkit request data is stored in payload field
    // https://a.yandex-team.ru/arc/trunk/arcadia/alice/vins/api/vins_api/speechkit/schemas.py?rev=5421968#L321
    // So this field is always empty
    optional TImageInputData Data = 6 [json_name = "data", (NYT.column_name) = "data"];

    // music_input
    optional TMusicResult MusicResult = 7 [json_name = "music_result", (NYT.column_name) = "music_result"];

    // server_action
    optional google.protobuf.Struct Payload = 8 [json_name = "payload", (NYT.column_name) = "payload", (NYT.flags) = SERIALIZATION_PROTOBUF];
    optional string Name = 9 [json_name = "name", (NYT.column_name) = "name"];
    optional bool IgnoreAnswer = 10 [json_name = "ignore_answer", (NYT.column_name) = "ignore_answer"];

    optional TBiometryScoring BiometryScoring = 11 [json_name = "biometry_scoring", (NYT.column_name) = "biometry_scoring"];
    optional TBiometryClassification BiometryClassification = 12 [json_name = "biometry_classification", (NYT.column_name) = "biometry_classification"];

    optional bool IsWarmUp = 13 [json_name = "is_warmup", (NYT.column_name) = "is_warmup"];
    // voice input (VOICESERV-3651)
    optional string AsrCoreDebug = 14 [json_name = "asr_core_debug", (NYT.column_name) = "asr_core_debug", (FieldAccess) = A_PRIVATE_REQUEST, (FieldAccess) = A_PRIVATE_RESPONSE];

    optional bool AsrWhisper = 16 [json_name = "asr_whisper", (NYT.column_name) = "asr_whisper"];
    optional uint64 AsrDurationProcessedAudio = 17 [json_name = "asr_duration_processed_audio", (NYT.column_name) = "asr_duration_processed_audio"];
};
//TODO: events
